src/ott/solvers/linear/univariate.py

```diff
@@ -12,134 +12,247 @@
-      :class:`PNormP(2) <ott.geometry.costs.PNormP>`.
-    method: The method used for computing the distance on the line. Options
-      currently supported are:
-
-      - `'subsample'` - Take a stratified sub-sample of the distances.
-      - `'quantile'` - Take equally spaced quantiles of the distances.
-      - `'equal'` - No subsampling is performed, requires distributions to have
-        the same number of points.
-      - `'wasserstein'` - Compute the distance using the explicit solution
-        involving inverse CDFs.
-
-    n_subsamples: The number of samples to draw for the "quantile" or
-      "subsample" methods.
+    num_subsamples: option to reduce the size of inputs by doing random
+      subsampling, taken into account marginal probabilities.
+    quantiles: When a vector or a number of quantiles is passed, the distance
+      is computed by evaluating the cost function on the sectional (one for each
+      dimension) quantiles of the two point cloud distributions described in the
+      problem.
   """
 
   def __init__(
       self,
-      sort_fn: Optional[Callable[[jnp.ndarray], jnp.ndarray]] = None,
-      cost_fn: Optional[costs.CostFn] = None,
-      method: Literal["subsample", "quantile", "wasserstein",
-                      "equal"] = "subsample",
-      n_subsamples: int = 100,
+      num_subsamples: Optional[int] = None,
+      quantiles: Optional[Union[int, jnp.ndarray]] = None,
   ):
-    self.sort_fn = jnp.sort if sort_fn is None else sort_fn
-    self.cost_fn = costs.PNormP(2) if cost_fn is None else cost_fn
-    self.method = method
-    self.n_subsamples = n_subsamples
+    self._quantiles = quantiles
+    self.num_subsamples = num_subsamples
+
+  @property
+  def quantiles(self):
+    """Quantiles' values used to evaluate OT cost."""
+    if self._quantiles is None:
+      return None
+    if isinstance(self._quantiles, int):
+      return jnp.linspace(0.0, 1.0, self._quantiles)
+    return self._quantiles
+
+  @property
+  def num_quantiles(self):
+    """Number of quantiles used to evaluate OT cost."""
+    return 0 if self._quantiles is None else self.quantiles.shape[0]
 
   def __call__(
       self,
-      x: jnp.ndarray,
-      y: jnp.ndarray,
-      a: Optional[jnp.ndarray] = None,
-      b: Optional[jnp.ndarray] = None
+      prob: linear_problem.LinearProblem,
+      rng: Optional[jax.Array] = None,
   ) -> float:
-    """Computes the Univariate OT Distance between `x` and `y`.
+    """Computes Univariate Distance between the `d` dimensional slices.
 
     Args:
-      x: The first distribution of shape ``[n,]`` or ``[n, 1]``.
-      y: The second distribution of shape ``[m,]`` or ``[m, 1]``.
-      a: The first marginals when ``method = 'wasserstein'``. If :obj:`None`,
-        uniform will be used.
-      b: The second marginals when ``method = 'wasserstein'``. If :obj:`None`,
-        uniform will be used.
+      prob: describing, in its geometry attribute, the two point clouds
```

Comment:
Would link more to the `LinearProblem`, e.g.:
```rst
Problem with a `point cloud <ott.geometry.pointcloud.PointCloud>` geometry
and a ground `TI cost <ott.geometry.costs.TICost>` between two scalars.
The ``[n,]`` and ``[m,]`` sized probability weights are stored in 
:attr:`~ott.problems.linear.LinearProblem.a` and ..., respectively.
```

Reply:
thanks so much for the great suggestion !!